[shard] remove set device for nccl

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69946

This PR remove the implicit set_device for nccl pg according to the proposal of https://github.com/pytorch/pytorch/issues/69731

Differential Revision: [D33099095](https://our.internmc.facebook.com/intern/diff/D33099095/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33099095/)!

Commit:
[shard] remove set device for nccl

This PR remove the implicit set_device for nccl pg according to the proposal of https://github.com/pytorch/pytorch/issues/69731

Differential Revision: [D33099095](https://our.internmc.facebook.com/intern/diff/D33099095/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33099095/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -271,13 +271,11 @@ def gather(
-                object_list=gathered_shards,
-                group=self._process_group,
-            )
+        dist.all_gather_object(
```
Can we also see if we can remove this one too: https://github.com/pytorch/pytorch/blob/master/torch/distributed/_sharded_tensor/__init__.py#L424

pritamdamania87 commented on torch/testing/_internal/distributed/_sharded_tensor/__init__.py:
```diff
@@ -27,6 +27,10 @@ def init_pg(self, backend="nccl"):
 
+        # set device for nccl pg for collectives
+        if backend == "nccl":
+            torch.cuda.set_device(self.rank)
```
We should update the docs of `ShardedTensor` to clearly mention this expectation from the user. We could actually even do an all_gather of all default device ids across all ranks of ShardedTensor and throw an error if the default device is the same on multiple ranks.

wanchaol replied:
updated the PR with the doc, quick question on the default device id, are you referring to that we can all_gather the `torch.cuda.current_device()` results from all rank and check if there's overlap?

wanchaol replied:
Let's add this in a follow up PR, I am actually a bit concerning on adding this, mainly because this means for every collective we run in nccl, we will need to run an additional all_gather just to check the default device overlap or not, which might potentially impact the perf?

wanchaol replied:
Created issue to track:
https://github.com/pytorch/pytorch/issues/70054
https://github.com/pytorch/pytorch/issues/70053

Commit:
Update on "[shard] remove set device for nccl"


This PR remove the implicit set_device for nccl pg according to the proposal of https://github.com/pytorch/pytorch/issues/69731

Differential Revision: [D33099095](https://our.internmc.facebook.com/intern/diff/D33099095/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33099095/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -258,6 +258,14 @@ def gather(
                 Must to be provided ONLY on ``dst`` rank.
                 Default: ``None``
+
+    .. note:: For NCCL-based processed groups, internal tensor representations
+        of objects must be moved to the GPU device before communication takes
+        place. In this case, the device used is given by
+        ``torch.cuda.current_device()`` and it is the user's responsiblity to
+        ensure that this is set so that each rank has an individual GPU, via
+        ``torch.cuda.set_device()``
+
```
I think we should put these docs in the ShardedTensor constructor docs. Essentially if you use ShardedTensor you should always do this irrespective of which API you use.

Users don't use the ShardedTensor constructor typically but we could have a sharded_tensor.rst mentioning this in our public docs in the future.

Commit:
Update on "[shard] remove set device for nccl"


This PR remove the implicit set_device for nccl pg according to the proposal of https://github.com/pytorch/pytorch/issues/69731

Differential Revision: [D33099095](https://our.internmc.facebook.com/intern/diff/D33099095/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33099095/)!

[ghstack-poisoned]
